dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/Period.java

```diff
@@ -425,4 +425,10 @@ public boolean isDefault()
     {
         return Objects.isNull( dateField );
     }
+
+    private Date atEndOfDay( Date date )
```

Comment:
Do we know for sure all dates passed in here has 00:00:00 time? Would it make sense to set time to 00:00:00 then add the offset?

Reply:
We do. I started doing this with java.time but then you have to consider time-zones and sort of rely on that all involved code has the same idea of time zone which is almost unlikely since timestamps are by definition UTC and system zone is unlike to be UTC. Therefore I just assumed whatever is passed as start is "zero time" so I hard added a milli less than a day. But the results were the same both ways. Lots of trouble. I think we have to give up on this way of fixing it.